src/main/java/com/team766/controllers/CanSparkMaxSmartMotionPIDController.java

```diff
@@ -0,0 +1,222 @@
+package com.team766.controllers;
+
+import com.team766.config.ConfigFileReader;
+import com.team766.hal.RobotProvider;
+import com.team766.library.SetValueProvider;
+import com.team766.library.SettableValueProvider;
+import com.team766.library.ValueProvider;
+import com.team766.logging.Category;
+import com.team766.logging.Logger;
+import com.team766.logging.Severity;
+
+
+import com.revrobotics.SparkMaxAbsoluteEncoder;
+import com.revrobotics.SparkMaxPIDController;
+import com.revrobotics.CANSparkMax.ControlType;
+import com.revrobotics.CANSparkMax.IdleMode;
+import com.revrobotics.SparkMaxAbsoluteEncoder.Type;
+import java.io.IOError;
+import javax.swing.text.DefaultStyledDocument.ElementSpec;
+import com.ctre.phoenix.motorcontrol.NeutralMode;
+import com.revrobotics.CANSparkMax;
+import com.team766.framework.Mechanism;
+import com.team766.hal.MotorController;
+import com.team766.library.RateLimiter;
+
+
+public class CanSparkMaxSmartMotionPIDController{
+	// The attributes of the class include references to the motor controller, SparkMax controller, PID controller, and absolute encoder
+	private MotorController mc;
+	private CANSparkMax csm;
+	private SparkMaxPIDController pid;
+	private SparkMaxAbsoluteEncoder abs;
+	//PID Related Variables
+	private static double dz1 = 0; 
+	private static double setPointPosition = 0;
+	private static double comboOfTimesInsideDeadzone = 0;
+	private static double minPos = 0;
+	private static double maxPos = 0;
+
+	private bool isAbsoluteEncoderEnabled;
+
+	//antigrav coefficient
+	private static double antiGravK;
+	//enum for which state the PID is in
+	public enum PIDSTATE{
+		PID,
+		OFF,
+		ANTIGRAV
+	}
+	//the state of the PID
+	private PIDSTATE theState = PIDSTATE.OFF;
+	
+	//constructor for the class with no absolute encoder
+	public CanSparkMaxSmartMotionPIDController(String configName) throws RuntimeException{
+			loggerCategory = Category.MECHANISMS;
+
+			try{
+				mc = RobotProvider.instance.getMotor(configName);
+				csm = (CANSparkMax)mc;
+				pid = csm.getPIDController();
+				isAbsoluteEncoderEnabled = true;
+			}catch (IllegalArgumentException ill){
+				throw new RuntimeException("Error instantiating the PID controller: " + ill);
+			}
+
+		
+	}
+	//constructor for the class with an absolute encoder
+	public CanSparkMaxSmartMotionPIDController(String configName, double absEncoderOffset) throws RuntimeException{
+			loggerCategory = Category.MECHANISMS;
+
+			try{
+				mc = RobotProvider.instance.getMotor(configName);
+				csm = (CANSparkMax)mc;
+				pid = csm.getPIDController();
+				abs = csm.getAbsoluteEncoder(Type.kDutyCycle);
+				abs.setZeroOffset(absEncoderOffset);
+				pid.setFeedbackDevice(abs);
```

Comment:
we should consider how we want to handle errors from Rev method calls. see `revErrorToException` and error handling in CANSparkMaxMotorController.java

Reply:
I think we can just allow the user to deal with them since it will be them using a value/something that isnt supported, so then they can see that and make adjustments

Reply:
i'm not sure how the user would deal with them, since the user is not informed that an error has occurred - either we would need to return the error code or throw an exception if an error occurred